tools/compiletest/src/main.rs

```diff
@@ -385,6 +386,46 @@ fn collect_expected_tests_from_dir(
     Ok(())
 }
 
+/// Collect `exec` tests from a directory.
+///
+/// Note that this method isn't recursive like
+/// `collect_expected_tests_from_dir`, as it allows us to ensure that newly
+/// added tests are running. Hence, each test must be in its own folder.
+fn collect_exec_tests_from_dir(
+    config: &Config,
+    dir: &Path,
+    relative_dir_path: &Path,
+    inputs: &Stamp,
+    tests: &mut Vec<test::TestDescAndFn>,
+) -> io::Result<()> {
+    // If we find a test `foo/bar.rs`, we have to build the
+    // output directory `$build/foo` so we can write
+    // `$build/foo/bar` into it. We do this *now* in this
+    // sequential loop because otherwise, if we do it in the
+    // tests themselves, they race for the privilege of
+    // creating the directories and sometimes fail randomly.
+    let build_dir = output_relative_path(config, relative_dir_path);
+    fs::create_dir_all(&build_dir).unwrap();
+
+    // Ensure the mode we're running is `Mode::Exec`
+    assert!(config.mode == Mode::Exec);
+
+    // Each test is expected to be in its own folder (i.e., we don't do recursion)
+    for file in fs::read_dir(dir)? {
+        // Look for `config.yml` file in folder
+        let file = file?;
+        let file_path = file.path();
+        let has_config_yml = file_path.join("config.yml").exists();
+        assert!(has_config_yml, "couldn't find `config.yml` file for `exec` test");
```

Comment:
Can we error out instead of panic?

Reply:
I tried using the `error!` macro, but this prints an error message (see 2nd line below) and continues, only to panic when executing the individual tests.

```
Check compiletest suite=exec-init mode=exec
2023-01-11T18:58:20.278416Z ERROR compiletest: couldn't find `config.yml` file for `exec` test

running 2 tests
F.
failures:

---- [exec] exec-init/error-code stdout ----
thread '[exec] exec-init/error-code' panicked at 'couldn't open `config.yml`: Os { code: 2, kind: NotFound, message: "No such file or directory" }', tools/compiletest/src/runtest.rs:333:57
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

In my opinion, it's better to have an early failure than waiting for the test to be run. Do you have a specific suggestion to replace the `assert!`?

Reply:
Yes, it's better to fail early. Perhaps use something similar to `fatal_proc_rec` in `runtest.rs`?